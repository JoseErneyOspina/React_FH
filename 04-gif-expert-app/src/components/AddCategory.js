import React, { useState } from 'react'
import PropTypes from 'prop-types';

const AddCategory = ( { setCategories } ) => {

    // const[inputValue, setInputValue] = useState(); // undefined
    const[inputValue, setInputValue] = useState(''); // ''

    const handleInputChange = ( e ) => {
        setInputValue( e.target.value );
    }

    const handleSubmit = ( e ) => {
        // Prevenimos el comportamiento por default
        e.preventDefault();
        if ( inputValue.trim().length > 2 ) {
            setCategories( cats => [ inputValue, ...cats ] );
            setInputValue('');
        }
    }

    return (
        <>
            <form onSubmit = { handleSubmit }>
                <h1> { inputValue } </h1>
                <input
                    type='text'
                    value = { inputValue }
                    onChange = { handleInputChange }
                />
            </form>
        </>
    )
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}

export default AddCategory;
