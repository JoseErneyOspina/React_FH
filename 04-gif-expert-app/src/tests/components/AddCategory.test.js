import React from 'react';
import { shallow } from 'enzyme';
import AddCategory from './../../components/AddCategory';

//ayuda
import '@testing-library/jest-dom';

describe('Testing at <AddCategory />', () => {

    const setCategories = jest.fn();
    let wrapper = shallow( <AddCategory setCategories = { setCategories } />);

    beforeEach( () => {
        jest.clearAllMocks();
        wrapper = shallow( <AddCategory setCategories = { setCategories } />);
    })

    test('Should show correctly', () => {
        expect( wrapper ).toMatchSnapshot();
    });

    test('Should change the text box', () => {
        const input = wrapper.fid('input');
        const value = 'Hola Mundo';
        input.simulate('change', { target: { value } });
        expect( wrapper.find('p').text().trim() ).toBe( value );
    });

    test('Should no post the info on submit', () => {
        wrapper.find('form').simulate('submit', { preventDefault(){} });
        expect( setCategories ).not.toHaveBeenCalled();
    });

    test('Should call the setCategories and clean the text box', () => {
        const value = 'Hola Mundo';
        // Simular el inputChange
        wrapper.fin('input').simulate('change', { target: { value } });
        // Simular el submit
        wrapper.fin('form').simulate('submit', { preventDefault(){} });

        // setCategories se debe de haber llamado
        expect( setCategories ).toHaveBeenCalled();
        // Ejemplo de cuantas veces se llamo
        expect( setCategories ).toHaveBeenCalledTimes(1);
        // Ejemplo de que se llame con una funci√≥n
        expect( setCategories ).toHaveBeenCalledWith( expect.any( Function ) );


        // El valor del input debe de estar ''
        expect( wrapper.find('input').prop('value') ).toBe('');
    });

});