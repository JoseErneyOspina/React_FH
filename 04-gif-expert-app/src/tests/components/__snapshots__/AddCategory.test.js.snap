// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing at <AddCategory /> Should show correctly 1`] = `
<Fragment>
  <form
    onSubmit={[Function]}
  >
    <p>
       
       
    </p>
    <input
      onChange={[Function]}
      type="text"
      value=""
    />
  </form>
</Fragment>
`;
